SingularityExecutor/src/main/java/com/hubspot/singularity/executor/config/SingularityExecutorTaskBuilder.java

```diff
@@ -83,6 +92,32 @@ public SingularityExecutorTask buildTask(String taskId, ExecutorDriver driver, T
     return new SingularityExecutorTask(driver, executorUtils, baseConfiguration, executorConfiguration, taskDefinition, executorPid, artifactFetcher, taskInfo, templateManager, log, jsonObjectFileHelper, dockerUtils, s3Configuration, jsonObjectMapper);
   }
 
+  private void checkAssignedPorts(TaskInfo taskInfo) {
+    for (Resource portsResource : taskInfo.getResourcesList().stream()
+        .filter((r) -> r.getName().equals("ports"))
+        .collect(Collectors.toList())) {
+      for (Range r : portsResource.getRanges().getRangeList()) {
+        for (long port = r.getBegin(); port <= r.getEnd(); port++) {
+          if (isPortInUse((int) port)) {
+            throw new RuntimeException(String.format("Assigned port %d was already in use", port));
+          }
+        }
+      }
+    }
+
+  }
+
+  private boolean isPortInUse(int port) {
+    try {
+      new Socket("127.0.0.1", port).close();
```

Comment:
You should create a ServerSocket on `0.0.0.0` to attempt to bind the port, rather than attempting to connect to it. There are a lot of reasons the connection to the port could fail, and it could even block for a bit trying to connect which would cause this function to hang.

Additionally, there's no real guarantee in singularity that the app will bind to 127.0.0.1, it could just bind to the main IP of the host and you'd miss it, so binding 0.0.0.0 would ensure it's not listening to the port on any address.

Reply:
good point, will switch to ServerSocket